File: app/src/test/java/com/kickstarter/viewmodels/ProjectUpdatesViewModelTest.java
package com.kickstarter.viewmodels;

import android.content.Intent;
import android.util.Pair;

import com.kickstarter.KSRobolectricTestCase;
import com.kickstarter.libs.Environment;
import com.kickstarter.mock.factories.ProjectFactory;
import com.kickstarter.mock.factories.UpdateFactory;
import com.kickstarter.mock.services.MockApiClient;
import com.kickstarter.models.Project;
import com.kickstarter.models.Update;
import com.kickstarter.services.apiresponses.UpdatesEnvelope;
import com.kickstarter.ui.IntentKey;

import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import androidx.annotation.NonNull;
import rx.Observable;
import rx.observers.TestSubscriber;

public class ProjectUpdatesViewModelTest extends KSRobolectricTestCase {
  private ProjectUpdatesViewModel.ViewModel vm;
  private final TestSubscriber<Boolean> isFetchingUpdates = new TestSubscriber<>();
  private final TestSubscriber<List<Update>> updates = new TestSubscriber<>();
  private final TestSubscriber<Pair<Project, Update>> startUpdateActivity = new TestSubscriber<>();

  private void setUpEnvironment(final @NonNull Environment env, final @NonNull Project project) {
    this.vm = new ProjectUpdatesViewModel.ViewModel(env);
    this.vm.outputs.isFetchingUpdates().subscribe(this.isFetchingUpdates);
    this.vm.outputs.startUpdateActivity().subscribe(this.startUpdateActivity);
    this.vm.outputs.updates().subscribe(this.updates);

    // Configure the view model with a project intent.
    this.vm.intent(new Intent().putExtra(IntentKey.PROJECT, project));
  }

  @Test
  public void testIsFetchingUpdates() {
    setUpEnvironment(environment(), ProjectFactory.project());

    this.isFetchingUpdates.assertValues(true, false);
    this.koalaTest.assertValue("Viewed Updates");
  }

  @Test
  public void testStartUpdateActivity() {
    final Update update = UpdateFactory.update();
    final List<Update> updates = Collections.singletonList(update);

    final Project project = ProjectFactory.project();
    setUpEnvironment(environment().toBuilder().apiClient(new MockApiClient() {
      @NonNull
      @Override
      public Observable<UpdatesEnvelope> fetchUpdates(@NonNull Project project) {
        return Observable.just(
          UpdatesEnvelope
            .builder()
            .updates(updates)
            .urls(urlsEnvelope())
            .build()
        );
      }
    }).build(), project);

    this.vm.inputs.updateClicked(update);

    this.startUpdateActivity.assertValues(Pair.create(project, update));
    this.koalaTest.assertValues("Viewed Updates", "Viewed Update");
  }

  @Test
  public void testUpdates() {
    final List<Update> updates = Arrays.asList(
      UpdateFactory.update(),
      UpdateFactory.update()
    );

    final Project project = ProjectFactory.project();
    setUpEnvironment(environment().toBuilder().apiClient(new MockApiClient() {
      @NonNull
      @Override
      public Observable<UpdatesEnvelope> fetchUpdates(@NonNull Project project) {
        return Observable.just(
          UpdatesEnvelope
            .builder()
            .updates(updates)
            .urls(urlsEnvelope())
            .build()
        );
      }
    }).build(), project);

    this.updates.assertValue(updates);
    this.koalaTest.assertValue("Viewed Updates");
  }

  private UpdatesEnvelope.UrlsEnvelope urlsEnvelope() {
    return UpdatesEnvelope.UrlsEnvelope
      .builder()
      .api(
        UpdatesEnvelope.UrlsEnvelope.ApiEnvelope
          .builder()
          .moreUpdates("http://more.updates.please")
          .build()
      )
      .build();
  }
}

File: app/src/test/java/com/kickstarter/viewmodels/ProjectUpdatesViewModelTest.kt
package com.kickstarter.viewmodels

import android.content.Intent
import android.util.Pair
import com.kickstarter.KSRobolectricTestCase
import com.kickstarter.libs.Environment
import com.kickstarter.libs.utils.EventName
import com.kickstarter.mock.factories.ProjectDataFactory.project
import com.kickstarter.mock.factories.ProjectFactory.project
import com.kickstarter.mock.factories.UpdateFactory.update
import com.kickstarter.mock.services.MockApolloClient
import com.kickstarter.models.Project
import com.kickstarter.models.Update
import com.kickstarter.services.apiresponses.UpdatesEnvelope
import com.kickstarter.services.apiresponses.updatesresponse.UpdatesGraphQlEnvelope
import com.kickstarter.ui.IntentKey
import com.kickstarter.ui.data.ProjectData
import org.junit.Test
import rx.Observable
import rx.observers.TestSubscriber
import rx.subjects.BehaviorSubject

class ProjectUpdatesViewModelTest : KSRobolectricTestCase() {
    private lateinit var vm: ProjectUpdatesViewModel.ViewModel

    private val horizontalProgressBarIsGone = TestSubscriber<Boolean>()
    private val isFetchingUpdates = TestSubscriber<Boolean>()
    private val projectAndUpdates = TestSubscriber<Pair<Project, List<Update>>>()
    private val startUpdateActivity = TestSubscriber<Pair<Project, Update>>()

    private fun setUpEnvironment(env: Environment, project: Project, projectData: ProjectData) {
        vm = ProjectUpdatesViewModel.ViewModel(env)
        vm.outputs.horizontalProgressBarIsGone().subscribe(horizontalProgressBarIsGone)
        vm.outputs.isFetchingUpdates().subscribe(isFetchingUpdates)
        vm.outputs.projectAndUpdates().subscribe(projectAndUpdates)
        vm.outputs.startUpdateActivity().subscribe(startUpdateActivity)

        // Configure the view model with a project intent.
        vm.intent(
            Intent().putExtra(IntentKey.PROJECT, project)
                .putExtra(IntentKey.PROJECT_DATA, projectData)
        )
    }

    @Test
    fun init_whenViewModelInstantiated_shouldTrackPageViewEvent() {
        val project = project()
        setUpEnvironment(environment(), project, project(project))

        segmentTrack.assertValue(EventName.PAGE_VIEWED.eventName)
    }

    @Test
    fun testHorizontalProgressBarIsGone() {
        val project = project()
        setUpEnvironment(environment(), project, project(project))

        horizontalProgressBarIsGone.assertValues(false, true)
    }

    @Test
    fun testIsFetchingUpdates() {
        val project = project()
        setUpEnvironment(environment(), project, project(project))

        isFetchingUpdates.assertValues(true, false)
    }

    @Test
    fun testProjectAndUpdates() {
        val updates = listOf(
            update(),
            update()
        )
        val project = project()

        setUpEnvironment(
            environment().toBuilder().apolloClient(object : MockApolloClient() {
                override fun getProjectUpdates(
                    slug: String,
                    cursor: String?,
                    limit: Int
                ): Observable<UpdatesGraphQlEnvelope> {
                    return Observable.just(
                        UpdatesGraphQlEnvelope
                            .builder()
                            .updates(updates)
                            .build()
                    )
                }
            }).build(),
            project, project(project)
        )

        val projectAndUpdates = BehaviorSubject.create<Pair<Project, List<Update>>>()
        vm.outputs.projectAndUpdates().subscribe(projectAndUpdates)

        assertEquals(project, projectAndUpdates.value.first)
        assertEquals(updates[0], projectAndUpdates.value.second[0])
    }

    @Test
    fun test_projectAndUpdates_whenUpdatesListIsEmpty() {
        val project = project()

        setUpEnvironment(
            environment().toBuilder().apolloClient(object : MockApolloClient() {
                override fun getProjectUpdates(
                    slug: String,
                    cursor: String?,
                    limit: Int
                ): Observable<UpdatesGraphQlEnvelope> {
                    return Observable.just(
                        UpdatesGraphQlEnvelope
                            .builder()
                            .updates(emptyList())
                            .build()
                    )
                }
            }).build(),
            project, project(project)
        )

        val projectAndUpdates = BehaviorSubject.create<Pair<Project, List<Update>>>()
        vm.outputs.projectAndUpdates().subscribe(projectAndUpdates)

        assertEquals(project, projectAndUpdates.value.first)
        assertTrue(projectAndUpdates.value.second.isEmpty())

        isFetchingUpdates.assertValues(true, false)
        horizontalProgressBarIsGone.assertValues(false, true)
    }

    @Test
    fun testStartUpdateActivity() {
        val update = update()
        val updates = listOf(update)
        val project = project()

        setUpEnvironment(
            environment().toBuilder().apolloClient(object : MockApolloClient() {
                override fun getProjectUpdates(
                    slug: String,
                    cursor: String?,
                    limit: Int
                ): Observable<UpdatesGraphQlEnvelope> {
                    return Observable.just(
                        UpdatesGraphQlEnvelope
                            .builder()
                            .updates(updates)
                            .build()
                    )
                }
            }).build(),
            project, project(project)
        )

        vm.inputs.updateClicked(update)
        startUpdateActivity.assertValues(Pair.create(project, update))
    }

    private fun urlsEnvelope(): UpdatesEnvelope.UrlsEnvelope {
        return UpdatesEnvelope.UrlsEnvelope
            .builder()
            .api(
                UpdatesEnvelope.UrlsEnvelope.ApiEnvelope
                    .builder()
                    .moreUpdates("http://more.updates.please")
                    .build()
            )
            .build()
    }
}
